File: Objective-C Sample Code/Chapter 9/Going Further/Top10/Top10 WatchKit App Extension/InterfaceController.m
//
//  InterfaceController.m
//  Top10 WatchKit App Extension
//
//  Created by Cory Bohon on 12/5/16.
//  Copyright © 2016 Cory Bohon. All rights reserved.
//

#import "InterfaceController.h"
#import "NetworkController.h"
#import "JSONItem.h"
#import "MainRow.h"

@interface InterfaceController()

@property (nonatomic, weak) IBOutlet WKInterfaceTable *table;
@property (nonatomic, weak) IBOutlet WKInterfaceGroup *backgroundColorGroup;
@property (nonatomic, strong) NSArray *jsonItems;

@end


@implementation InterfaceController

- (void)awakeWithContext:(id)context
{
    [super awakeWithContext:context];
    [self reloadData];
    
    [self animateWithDuration:2.0f animations:^{
        [self.backgroundColorGroup setBackgroundColor:[UIColor blackColor]];
    }];
}

#pragma mark - Table Methods
- (void)reloadData
{
    [[NetworkController sharedNetworkController] retrieveJSONFeedWithCompletionHandler:^(NSError *error, NSArray *objects) {
        
        dispatch_async(dispatch_get_main_queue(), ^{
            [self.table insertRowsAtIndexes:[NSIndexSet indexSetWithIndexesInRange:NSMakeRange(0, 10)] withRowType:@"MainRow"];
            
            for (NSUInteger index = 0; index < self.table.numberOfRows; index ++) {
                MainRow *row = [self.table rowControllerAtIndex:index];
                
                [row configureWithJSONItem:[objects objectAtIndex:index] atIndex:index];
                
                [row.image setHidden:NO];
                [row.label setHidden:NO];
            }
        });
        
        self.jsonItems = objects;
        
    } forNumberOfItems:10];
}

- (void)table:(WKInterfaceTable *)table didSelectRowAtIndex:(NSInteger)rowIndex
{
    [[WKInterfaceDevice currentDevice] playHaptic:WKHapticTypeClick];
    JSONItem *jsonItem = [self.jsonItems objectAtIndex:rowIndex];
    [self pushControllerWithName:@"detailController" context:jsonItem];
}

#pragma mark - Actions
- (IBAction)reloadMenuButtonTapped:(id)sender
{
    [self reloadData];
}

@end

File: Swift Sample Code/Chapter 12/Going Further/Top10/Top 10 WatchKit App Extension/InterfaceController.swift
//
//  InterfaceController.swift
//  Top 10 WatchKit App Extension
//
//  Created by Cory Bohon on 2/13/17.
//  Copyright © 2017 Cory Bohon. All rights reserved.
//

import WatchKit
import Foundation


class InterfaceController: WKInterfaceController {

    @IBOutlet weak var table: WKInterfaceTable?
    @IBOutlet weak var backgroundGroup: WKInterfaceGroup?
    
    var jsonItems: Array<Any>?
    
    override func awake(withContext context: Any?) {
        super.awake(withContext: context)
        self.setTitle("Top 10")
        
        self.reloadData()
    }
    
    override func didAppear() {
        super.didAppear()
        self.backgroundGroup?.setBackgroundColor(.black)
        
        self.updateUserActivity("com.Top10.website", userInfo: nil, webpageURL: URL(string: "https://www.apple.com/itunes/charts/songs/"))
    }
    
    @IBAction func reloadData() {
        NetworkController.shared().retrieveJSONFeed(completionHandler: { (error, objects) in
            
            DispatchQueue.main.async {
                
                if error != nil {
                    self.presentAlert()
                }
                
                self.table?.insertRows(at: IndexSet(integersIn: 0..<10), withRowType: "MainRow")
                
                for row in 0..<10 {
                    let currentRow: MainRow = self.table?.rowController(at: row) as! MainRow
                    currentRow.configure(item: objects[row] as! JSONItem, index: row)
                    
                    currentRow.image?.setHidden(false)
                    currentRow.label?.setHidden(false)
                }
            }
            
            self.jsonItems = objects
            
        }, forNumberOfItems: 10)
    }
    
    override func table(_ table: WKInterfaceTable, didSelectRowAt rowIndex: Int) {
        
        WKInterfaceDevice.current().play(.click)
        
        let jsonItem: JSONItem = self.jsonItems![rowIndex] as! JSONItem
        
        self.pushController(withName: "detailController", context: jsonItem)
    }
    
    func presentAlert() {
        let cancelAction: WKAlertAction = WKAlertAction(title: "Cancel", style: .cancel) { 
            // no handler needed
        }
        
        let retryAction: WKAlertAction = WKAlertAction(title: "Retry", style: .default) { 
            self.reloadData()
        }
        
        self.presentAlert(withTitle: nil, message: "Top 10 list currently unavailable", preferredStyle: .sideBySideButtonsAlert, actions: [cancelAction, retryAction])
    }
}
